mqtt/mqtt-bridge/src/persist/waking_state/mod.rs

```diff
@@ -0,0 +1,356 @@
+#![allow(dead_code)] // TODO remove when ready
+
+use std::{
+    cmp::min, collections::hash_map::Entry, collections::HashMap, collections::VecDeque,
+    task::Waker,
+};
+
+use async_trait::async_trait;
+use bincode::{self};
+use mqtt3::proto::Publication;
+use rocksdb::{IteratorMode, Options, DB};
+use uuid::Uuid;
+
+use crate::persist::{Key, PersistError};
+
+/// Responsible for waking waiting streams when new elements are added.
+///
+/// Exposes a get method for retrieving a count of elements in order of insertion.
+#[async_trait]
+pub trait StreamWakeableState {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError>;
+
+    fn batch(&mut self, count: usize) -> Result<Vec<(Key, Publication)>, PersistError>;
+
+    fn remove_in_flight(&mut self, key: &Key) -> Result<Publication, PersistError>;
```

Comment:
I converted this interface to return result in all cases. Previously we aligned with standard collection types and returned Option. I feel like being consistent with error is the best choice because we should never attempt to remove anything that doesn't exist.